src/persist-client/src/internal/encoding.rs

```diff
@@ -1144,4 +1189,85 @@ mod tests {
         };
         assert_eq!(<WriterState<u64>>::from_proto(proto).unwrap(), expected);
     }
+
+    #[test]
+    fn state_migration_rollups() {
+        let r1 = HollowRollup {
+            key: PartialRollupKey("foo".to_owned()),
+            encoded_size_bytes: None,
+        };
+        let r2 = HollowRollup {
+            key: PartialRollupKey("bar".to_owned()),
+            encoded_size_bytes: Some(2),
+        };
+        let shard_id = ShardId::new();
+        let mut state = TypedState::<(), (), u64, i64>::new(
+            DUMMY_BUILD_INFO.semver_version(),
+            shard_id,
+            "host".to_owned(),
+            0,
+        );
+        state.state.collections.rollups.insert(SeqNo(2), r2.clone());
+        let mut proto = state.into_proto();
+
+        // Manually add the old rollup encoding.
+        proto.deprecated_rollups.insert(1, r1.key.0.clone());
+
+        let state = UntypedState::<u64>::from_proto(proto).unwrap();
+        let state = state.check_codecs::<(), (), i64>(&shard_id).unwrap();
+        let expected = vec![(SeqNo(1), r1), (SeqNo(2), r2)];
+        assert_eq!(
+            state
+                .state
+                .collections
+                .rollups
+                .into_iter()
+                .collect::<Vec<_>>(),
+            expected
+        );
+    }
+
+    #[test]
+    fn state_diff_migration_rollups() {
+        let r1key = PartialRollupKey("foo".to_owned());
+        let r1 = StateFieldDiff {
+            key: SeqNo(1),
+            val: StateFieldValDiff::Insert(HollowRollup {
+                key: r1key.clone(),
+                encoded_size_bytes: None,
+            }),
+        };
+        let r2 = StateFieldDiff {
+            key: SeqNo(2),
+            val: StateFieldValDiff::Insert(HollowRollup {
+                key: PartialRollupKey("bar".to_owned()),
+                encoded_size_bytes: Some(2),
+            }),
+        };
+        let mut diff = StateDiff::<u64>::new(
+            DUMMY_BUILD_INFO.semver_version(),
+            SeqNo(3),
+            SeqNo(4),
+            0,
+            PartialRollupKey("ignored".to_owned()),
+        );
+        diff.rollups.push(r2.clone());
+        let mut proto = diff.into_proto();
+
+        // Manually add the old rollup encoding.
+        field_diffs_into_proto(
+            ProtoStateField::DeprecatedRollups,
+            &[StateFieldDiff {
+                key: r1.key,
+                val: StateFieldValDiff::Insert(r1key),
+            }],
+            proto.field_diffs.as_mut().unwrap(),
+            |k| k.into_proto().encode_to_vec(),
+            |v| v.into_proto().encode_to_vec(),
+        );
+
+        let diff = StateDiff::<u64>::from_proto(proto).unwrap();
+        let expected = vec![r2, r1];
+        assert_eq!(diff.rollups.into_iter().collect::<Vec<_>>(), expected);
```

Comment:
I wouldn't mind a follow-up check here that if we send a Delete diff to `ProtoStateField::Rollups` for `r1.key` we remove the diff, to show the lifecycle of reading in a rollup from the deprecated field and still being able to delete it via diff applied to the new field

Reply:
fantastic idea and definitely worth eating a CI cycle. done!

(I actually had a bug in my original impl here. it did end up getting caught by the tests, but it spooked me enough that I'm being extra careful. I actually re-introduced some backward compatibility bugs on purpose and made sure that the [normal CI upgrade tests caught them](https://buildkite.com/materialize/tests/builds?branch=danhhz%3Apersist_rollup_citest) (they did, 👏 philip 👏 ))